### ISSUE TITLE
Classes and Modules

### ISSUE THREAD

**Issue by kbparagua** [kbparagua/ruby-styleguide#16]

**Comment by kbparagua** [comment]
## Line Breaks

- **Option A** - 👍

- **Option B** - 😄

**Comment by kbparagua** [comment]
## Visibility Modifiers

- **Option A** - 👍 

  Align modifier with class declaration.
  ```ruby
  class MyClass
    def my_public_method
      # do something
    end

  private
    
    def my_helper_method
      # do something
    end
  end
  ```
- **Option B** - 😄 

  Align modifier with method declarations.
  ```ruby
  class MyClass
    def my_public_method
      # do something
    end

    private
    
    def my_helper_method
      # do something
    end
  end
  ```

**Comment by kbparagua** [comment]
## `self`

- **Option A** - 👍 

  Use `self` inside methods.
  ```ruby
  class MyClass
    def foo(x)
      self.x + self.y + self.z * x
    end
  end
  ```

- **Option B** - 😄 

  Do not use `self` inside methods, except when a local variable is shadowing a class variable/method.
  ```ruby
  class MyClass
    def foo(x)
      self.x + y + z * x
    end
  end
  ```

**Comment by b-ggs** [comment]
Suggestion for class methods:

```
class << self
  def class_method
    ...
  end
end
```

vs.

```
def self.class_method
  ...
end
```

**Comment by kbparagua** [comment]
## Class Methods

- **Option A** - 👍 
  ```ruby
  class Foo
    def self.my_method
      # do something
    end
  end
  ```

- **Option B** - 😄 
  ```ruby
  class Foo
    class << self
      def my_method
        # do something
      end
    end
  end
  ```

**Comment by hjensen1** [comment]
Can anyone provide reasoning for making a rule for `self.my_method` over `class << self`? I agree that `self.my_method` is more clear in general, but `class << self` seems better in classes with mostly/only class methods.

**Comment by b-ggs** [comment]
@hjensen1 Personally, it's just that I immediately know just from looking at the method block that it's a class method.

Without the `self.`, the method block could be easily mistaken for an instance method if the `class << self` is outside the current view. (Which was a mistake I've made more than once while looking at one of our models.)

**Comment by maynardlandrito** [comment]
@hjensen1 I agree with @b-ggs . This is specially true for classes with a lot of class and instance methods. If a class only has class methods or majority class methods, I guess `class << self` would be ok, but then we'll still have 2 different ways of doing it.
